hcache/tdb.c

```diff
@@ -0,0 +1,135 @@
+/**
+ * @file
+ * TDB backend for the header cache
+ *
+ * @authors
+ * Copyright (C) 2020 Tino Reichardt <milky-neomutt@mcmilk.de>
+ *
+ * @copyright
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 2 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @page hc_tdb TDB
+ *
+ * Use a GNU dbm file as a header cache backend.
+ */
+
+#include "config.h"
+#include <stddef.h>
+#include "mutt/lib.h"
+#include "backend.h"
+#include "globals.h"
+
+#include <fcntl.h>
+#include <tdb.h>
+
+/**
+ * hcache_tdb_open - Implements HcacheOps::open()
+ */
+static void *hcache_tdb_open(const char *path)
+{
+  int flags = TDB_NOLOCK|TDB_INCOMPATIBLE_HASH|TDB_NOSYNC;
+  return tdb_open(path, 33657, flags, O_CREAT|O_RDWR, 00600);
+}
+
+/**
+ * hcache_tdb_fetch - Implements HcacheOps::fetch()
+ */
+static void *hcache_tdb_fetch(void *ctx, const char *key, size_t keylen, size_t *dlen)
+{
+  if (!ctx)
+    return NULL;
+
+  TDB_CONTEXT *db = ctx;
+  TDB_DATA dkey;
+  TDB_DATA data;
+
+  dkey.dptr = (unsigned char *) key;
+  dkey.dsize = keylen;
+  data = tdb_fetch(db, dkey);
+
+  *dlen = data.dsize;
+  return data.dptr;
+}
+
+/**
+ * hcache_tdb_free - Implements HcacheOps::free()
+ */
+static void hcache_tdb_free(void *vctx, void **data)
+{
+  FREE(data);
+}
+
+/**
+ * hcache_tdb_store - Implements HcacheOps::store()
+ */
+static int hcache_tdb_store(void *ctx, const char *key, size_t keylen, void *data, size_t dlen)
+{
+  if (!ctx)
+    return -1;
+
+  TDB_DATA dkey;
+  TDB_DATA databuf;
```

Comment:
Elsewhere you call this just ‘data’

Reply:
data is the parameter of that function :/

Reply:
Oh, sure... nevermind